File: merge_sort.rb

def merge_sort(array)
  array_length = array.length
  return array if array_length <= 1

  first_half = merge_sort(array[0...array_length/2])
  second_half = merge_sort(array[array_length/2..array_length])

  sorted_array = []
  pointer1, pointer2 = 0, 0

  (0).upto(array.length-1) do |i|
    if first_half[pointer1].nil?
      sorted_array[i] = second_half[pointer2]
      pointer2 += 1
    elsif second_half[pointer2].nil?
      sorted_array[i] = first_half[pointer1]
      pointer1 += 1
    elsif first_half[pointer1] < second_half[pointer2]
      sorted_array[i] = first_half[pointer1]
      pointer1 += 1
    else
      sorted_array[i] = second_half[pointer2]
      pointer2 += 1
    end
  end
  sorted_array
end

puts merge_sort([]) == []
puts merge_sort([1]) == [1]
puts merge_sort([1,1]) == [1,1]
puts merge_sort([2,1,2]) == [1,2,2]
puts merge_sort([5,4,1,8,7,2,6,3]) == [1,2,3,4,5,6,7,8]

File: javascript/merge_sort.js
array = {};

var array = [0,1,2,3,4,5];

function mergeSort (array) {
   if (array.length <= 1){
    return array;
   }
   var midpoint = Math.floor(array.length/2);
   var first_half = mergeSort(array.slice(0,midpoint));
   var second_half = mergeSort(array.slice(midpoint,array.length+1));
   return merge(first_half, second_half);
}

function merge (array1, array2) {
  var iterations = array1.length + array2.length;
  var pointer1 = 0;
  var pointer2 = 0;
  var sorted = [];
  for (var i = 0; i < iterations; i++){
    if (array1[pointer1] !== undefined &&
        array2[pointer2] !== undefined &&
        array1[pointer1] <= array2[pointer2]){
        sorted.push(array1[pointer1]);
        pointer1++;
      } else if (array1[pointer1] !== undefined && array2[pointer2] !== undefined){
        sorted.push(array2[pointer2]);
        pointer2++;
      } else if (array1[pointer1] !== undefined) {
        sorted.push(array1[pointer1]);
        pointer1++;
      } else {
        sorted.push(array2[pointer2]);
        pointer2++;
    }
  }
  return sorted;
}


assert(mergeSort([1,3,5,7,4,6,9,8,2,10]).compare([1,2,3,4,5,6,7,8,9,10]), "mergeSort passed");
